### PR TITLE
Reduce draw calls

### LINKED ISSUE #6544: Lowering draw calls
Continue applying the ideas and tools described in #6001 to improve rendering performance by reducing the number of distinct draw calls.

### LINKED ISSUE #6717: New shape instance render latency
### Discussed in https://github.com/orgs/enso-org/discussions/6716

<div type='discussions-op-text'>

<sup>Originally posted by **kazcw** May 16, 2023</sup>
When I create a new instance of a shape, I'm observing a 1-frame delay before it starts drawing. This is not caused by GPU buffer upload; other GPU parameter changes are visible without the delay. I suspect it is related to asynchronous processing of changes in the display object hierarchy. Is this expected behavior? It is causing a flicker in the new edge implementation when I need to change shaders in response to hover position changes (#6707), and working around it by delaying hiding the previous shader would add significant complexity.</div>

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Implements #6544 (eliminates 10/42 of the constantly-displayed draw calls).
Fixes #6717. Improves startup CPU time by 5% (250ms, loading Orders on my dev box).

### Important Notes

- Edges: New implementation uses only Rectangle under most conditions.
- Node and action area: Replace some shapes with Rectangle.
- List view: Replace some shapes with Rectangle.
- Display object hierarchy: The lowest-level shape instance types no longer have their own display objects.
- Includes initial support for using `Rectangle` to display triangles.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by kazcw** [comment]
I'll be testing this some more in the morning (with a Book Club), but it's ready for review.

File: app/gui/view/graph-editor/src/component/edge.rs [hunk 1]
```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
...
+        let mouse_move = display_object.on_event::<mouse::Move>();
+        let mouse_down = display_object.on_event::<mouse::Down>();
+        let mouse_out = display_object.on_event::<mouse::Out>();
 
-            fn set_color(&self, color:color::Rgba) {
-                self.color_rgba.set(Vector4(color.red,color.green,color.blue,color.alpha));
-            }
+        let output = &frp.private.output;
+        frp::extend! { network
+            // Setters. Changes to these values don't automatically trigger a redraw.
```
Comment (vitvakatu) [inline app/gui/view/graph-editor/src/component/edge.rs]: I think this should be mentioned in the documentation of the FPR definition.

File: app/gui/view/graph-editor/src/component/edge.rs [hunk 2]
```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
...
+            eval frp.target_position ((t) model.set_target_position(*t));
+            eval frp.source_attached ((t) model.set_source_attached(*t));
+            eval frp.target_attached ((t) model.set_target_attached(*t));
+            eval frp.source_width    ((t) model.set_source_width(*t));
+            eval frp.source_height   ((t) model.set_source_height(*t));
+            eval frp.set_disabled    ((t) model.set_disabled(*t));
+
+            // Mouse events.
+            eval mouse_move            ((e) model.set_mouse_position_and_redraw(e.client_centered()));
```
Comment (vitvakatu) [inline app/gui/view/graph-editor/src/component/edge.rs]: We no longer use spacing in FRP code, and it isn't consistent here anyway, so I suggest removing redundant spaces.

File: app/gui/view/graph-editor/src/component/edge.rs [hunk 3]
```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
...
+            hover_sections: default(),
+            corner_points: default(),
+            target_attachment: default(),
+            hover_position: default(),
+            disabled: default(),
+            previous_target: default(),
+            previous_is_hoverable: default(),
+            previous_hover_split: default(),
+            previous_color: default(),
+            previous_target_attached: default(),
```
Comment (vitvakatu) [inline app/gui/view/graph-editor/src/component/edge.rs]: I wonder if using `..default()` is reasonable here.

File: app/gui/view/graph-editor/src/component/edge.rs [hunk 4]
```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
...
+    }
 
-        impl AnyEdgeShape for $name {
-            #[allow(clippy::vec_init_then_push)]
-            fn shapes(&self) -> Vec<&dyn EdgeShape> {
-                let mut output = Vec::<&dyn EdgeShape>::default();
-                $(output.push(&self.$field);)*
-                output
-            }
+    fn source_x_range(&self) -> f32 {
```
Comment (vitvakatu) [inline app/gui/view/graph-editor/src/component/edge.rs]: What is `source_x_range`? I'd expect `Range<f32>` as the return type for anything with `range` in the name.

**Review by vitvakatu — COMMENTED** [review]
I can also see some changes to the list view. Please mention them in the PR description so QA reviewers know what to check carefully.

File: app/gui/view/graph-editor/src/component/edge.rs (hunk 3 above)
Comment (kazcw) [reply app/gui/view/graph-editor/src/component/edge.rs]: I wish I could do it that way, but I can't `impl Default for EdgeModel` because it must hold a `Scene`, so I can't default-construct an `EdgeModel` object to use for field-update syntax.

**Comment by kazcw** [comment]
There's a flicker when the mouse enters/leaves the arc area of a corner, caused by an existing bug. Issue: #6717

File: app/gui/view/graph-editor/src/component/edge.rs [hunk 5]
```diff
@@ -4,2033 +4,1044 @@ use crate::prelude::*;
...
+            // The point is along the vertical line. Snap its x-value, and draw a corner to it.
+            let snapped = Vector2(self.vertical.x(), split_point.y());
+            let source_end = self.with_target_end(snapped);
+            let target_end = self.with_source_end(snapped);
+            Some(SplitCorner { source_end, target_end, split_arc: None })
+        } else if on_arc {
+            // Find the input point's angle along the arc.
+            let offset_from_origin = split_point - origin;
+            let split_angle = offset_from_origin.y().atan2(offset_from_origin.x());
+            // Split the arc on the angle.
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/edge.rs]: In the new implementation, a lot of docs from the old implementation are missing, something like this:

```
// node (for example, not to draw the port line above source node).
         //
         // ╭──────────────╮
         // │      ┼ (0,0) │
         // ╰──────────────╯────╮
         //                     │
         //                     ▢ target
```

These docs were very helpful to understand which part of a connection the code relates to. Can we bring them back, please?

File: app/gui/view/graph-editor/src/component/edge.rs [hunk 6]
```diff
@@ -4,2033 +4,1044 @@ use crate::prelude::*;
...
+        self.redraw();
+    }
 
-                fn set_focus_split_angle(&self, angle: f32) {
-                    self.focus_split_angle.set(angle);
-                }
+    fn new_section(&self) -> Rectangle {
+        let new = Rectangle::new();
+        new.set_corner_radius_max();
+        new.set_inset_border(LINE_WIDTH);
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/edge.rs]: These functions should have at least small docs. Based on this name, it's hard to understand what a "section" is.

File: app/gui/view/graph-editor/src/component/edge.rs (hunk 3 above)
Comment (wdanilo) [reply app/gui/view/graph-editor/src/component/edge.rs]: We could split this into two structs and deref to the inner struct, making the inner struct Default. I'm not sure though if we want it. Up to you :)

**Review by wdanilo — CHANGES_REQUESTED** [review]
This is a GIGANTIC change, amazing job! I'm marking this as "request changes" only because:

1. There are missing docs in several places for private functions, whose names are not explanatory enough. Edge drawing is super complex and good docs there will be a big time saver for the future refactoring / improvements.
2. There is no video attached to this PR. If a PR touches visual things, a video showing as many edge cases as possible should be attached.


Also, Kaz, do you think you'd be able to create a debug scene with an edge only? (no nodes). Just an edge between two boxes, where you can drag one of the box? This will be insanely helpful for edge debugging / improving. I always wanted to do it, but it required getting into the edges drawing code - as you know this code well now, this should be doable pretty fast, what do you think about it? :)

**Comment by wdanilo** [comment]
Also, Kaz, could you share some stats, please? What is the draw call count drop after this PR? :)

**Comment by kazcw** [comment]
> Also, Kaz, do you think you'd be able to create a debug scene with an edge only? (no nodes). Just an edge between two boxes, where you can drag one of the box? This will be insanely helpful for edge debugging / improving. I always wanted to do it, but it required getting into the edges drawing code - as you know this code well now, this should be doable pretty fast, what do you think about it? :)

Created a `graph_view_edges` debug scene, that shows an edge between draggable dummy nodes. I also cleaned up the FRP *a lot*, which simplified the debug scene and graph-editor.

**Comment by kazcw** [comment]
> Also, Kaz, could you share some stats, please? What is the draw call count drop after this PR? :)

Testing Orders, showing just the graph right after opening:
Before (current `develop`):
image
After (this branch, currently):
image

Eliminated 10/42 (~24%) of the constantly-displayed draw calls.

File: app/gui/view/graph-editor/src/component/edge.rs (hunk 5 above)
Comment (kazcw) [reply app/gui/view/graph-editor/src/component/edge.rs]: Added docs with diagrams to the `layout` module, explaining the key concepts of the design.

**Comment by kazcw** [comment]
> 1. There are missing docs in several places for private functions, whose names are not explanatory enough. Edge drawing is super complex and good docs there will be a big time saver for the future refactoring / improvements.
> 2. There is no video attached to this PR. If a PR touches visual things, a video showing as many edge cases as possible should be attached.

Added a video, and a lot more docs. I think I've addressed everything. I also refactored the code a lot, and I think it's easier to read now.

**Review by wdanilo — CHANGES_REQUESTED** [review]
Based on the video, I've spotted several regressions:

1. The lower part of the edge is not above the nodes anymore:
    image
2. Edges going up do not have up arrow visible.
3. Edges have different corners than currently. The current behavior of edges was carefully designed to not inference with nodes, the new one is way more curvy and less uniform. Consider the old vs the new one:
     image 
    image
4. I think the behavior of edges close to node edge is different than it was. It was smoothly moving around and the corner radius was depending on the mouse distance:
https://github.com/enso-org/enso/assets/1623053/424ec53a-a2e1-4362-8f32-640c0cc0a781


Basically, we should preserve the look of edges as we have right now because we spent really lot of time designing their current behavior. If we want to change any part of it, let's have a call about it and let's talk about why such a change will introduce better UX to the user. But until we change the design, we need to preserve 100% of the look and feel when doing performance improvements.

**Review by wdanilo — CHANGES_REQUESTED** [review]
Based on the video, I've spotted several regressions:

1. The lower part of the edge is not above the nodes anymore:
    image
2. Edges going up do not have up arrow visible.
3. Edges have different corners than currently. The current behavior of edges was carefully designed to not inference with nodes, the new one is way more curvy and less uniform. Consider the old vs the new one:
     image 
    image
4. I think the behavior of edges close to node edge is different than it was. It was smoothly moving around and the corner radius was depending on the mouse distance:
https://github.com/enso-org/enso/assets/1623053/424ec53a-a2e1-4362-8f32-640c0cc0a781


Basically, we should preserve the look of edges as we have right now because we spent really lot of time designing their current behavior. If we want to change any part of it, let's have a call about it and let's talk about why such a change will introduce better UX to the user. But until we change the design, we need to preserve 100% of the look and feel when doing performance improvements.

**Comment by kazcw** [comment]
Updated demo; the old edges are in red:

[vokoscreenNG-2023-06-01_16-37-03.webm](https://github.com/enso-org/enso/assets/1047859/c958b522-13ab-457a-9d16-186a32a9ffc0)

Known differences:
- At the beginning of the video, you can see the radius is different in some positions when the target is very close to the source. The old edge snaps abruptly between a sharp corner and a smoother angle. The new edge interpolates between these extremes, avoiding a sudden transition in this case.
- Near the end of the video, when the target is slightly higher than the source, and barely to the side of the source, the old edges form a shark fin shape, with two different radii side by side. The new edges maintain symmetry in this case.

File: app/gui/view/graph-editor/src/component/edge/layout.rs [hunk 7]
```diff
@@ -0,0 +1,671 @@
...
+//!
+//! # Partial edges
+//!
+//! Corners are sufficient to draw any complete edge; however, in order to split an edge into a
+//! focused portion and an unfocused portion at an arbitrary location based on the mouse position,
+//! we need to subdivide one of the corners of the edge.
+//!
+//!                  |\
+//!                  | 3
+//!                 /
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/component/edge/layout.rs]: these docs are beautiful ❤️

**Review by wdanilo — APPROVED** [review]
Looks fantastic! Let's remove the old edge impl and let's merge it! Amazing job with fixing the corner cases where the old edges misbehaved!

**Comment by vitvakatu** [comment]
QA: passed 📗
